Fix part of #17712: Add curriculum admin topic management acceptance tests

Vir-8 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #17712.
2. This PR does the following: Adds acceptances tests for topic management, for the curriculum admin role. Tests the ability of the curriculum admin to create topics, subtopics, skills, stories, and chapters, as well as publish these creations.

## Essential Checklist

- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [X] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [X] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [X] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. (For changes involving responsiveness or adjustment
of UI elements, this should be a video that gradually resizes the viewport from wide
to narrow and back again.) If this PR is for a developer-facing feature, provide
the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be sure
that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->
<img width="568" alt="image" src="https://github.com/oppia/oppia/assets/73329256/918029be-d74d-46d1-a6db-e700f55b5af5">

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Vir-8 committed:
Create acceptance test files and implement functionality to add skills

Vir-8 committed:
Add functionality to create questions

Vir-8 committed:
Add functionality to create explorations

Vir-8 committed:
Add thumbnail image and topic creation functionality

Vir-8 committed:
Add functionality to create subtopics, stories, chapters, and publish stories

Vir-8 committed:
Write the acceptance tests to create topics, stories, chapters, and publish stories

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Migrate curriculum admin tests to TypeScript

Vir-8 committed:
Add comments and fix tests

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix linting issues

Vir-8 committed:
Fix utility functions

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix tests by adding timeouts for improved stability

Vir-8 committed:
Clean code

Vir-8 committed:
Fix typescript issues and set the main page to the active tab

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Change comments for clarity

Vir-8 committed:
Add expectation tests

Vir-8 committed:
Fix linting issues

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 commented:
@seanlip PTAL. Note that I had already added and tested the utility to create questions before I noticed it wasn't in scope of this test - didn't remove the utility as I'm sure it will save time and be useful in the near future when associated tests are written. Hope that works.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/curriculum-admin-tests/create-and-publish-topics-and-stories.spec.ts:
```diff
@@ -0,0 +1,65 @@
+    async function () {
+      await curriculumAdmin.navigateToCreatorDashboardPage();
+      explorationUrl = await curriculumAdmin.createExploration();
+      explorationId = explorationUrl
```
This is too complex for the top layer; extract it into a helper function.

Also I am a bit concerned about the default case. Shouldn't some error arise if an exploration ID cannot be extracted?

Vir-8 replied:
Made it simpler for the top layer, and by default, the acceptance test would fail during chapter creation - where it tries to use the ID. Modified it to throw an error if the ID isn't extracted.

seanlip replied:
That sounds good, thanks. (In general, fail as early as possible.)

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+const createTopicButton = 'button.e2e-test-confirm-topic-creation-button';
+const saveTopicButton = 'button.e2e-test-save-topic-button';
+
+const addSubTopicButton = 'button.e2e-test-add-subtopic-button';
```
Subtopic is one word, here and elsewhere

Vir-8 replied:
Done

seanlip replied:
@Vir-8 Just a note, please try not to resolve comments. Otherwise the reviewer needs to open each of them one by one to check their status. Thanks!

Vir-8 replied:
My bad, I thought it would help improve readability of what is more important. Really sorry!

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+    );
+    await this.clickOn(publishExplorationButton);
+    await this.type(explorationTitleInput, 'Test Exploration');
+    await this.type(explorationGoalInput, 'Test Exploration');
```
I suggest you write different things for the different strings, so that distinguishing them later (e.g. if we want to verify things) becomes easier in the future.

Vir-8 replied:
Overlooked that, done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+    await this.clickOn(addInteractionButton);
+    await this.clickOn(interactionEndExplorationInputButton);
+    await this.clickOn(saveInteractionButton);
+    await this.page.waitForSelector(`${saveChangesButton}:not([disabled])`);
```
Think it's worth incorporating this check in this.clickOn()? It kind of interrupts the description of the flow.

Vir-8 replied:
That check was redundant, removed it, thanks

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+  /**
+   * Function for creating a subtopic as a curriculum admin.
+   */
+  async createSubTopic(): Promise<void> {
```
Subtopic is one word

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+    await this.page.waitForSelector(`${uploadPhotoButton}:not([disabled])`);
+    await this.clickOn(uploadPhotoButton);
+    await this.page.waitForSelector(`${createChapterButton}:not([disabled])`);
+    await this.page.waitForTimeout(500);
```
Can we incorporate these waits in their respective actions (click, upload, etc.)? 

Also in general we try to avoid waiting for a timeout. This leads to race conditions since we don't know if the action will happen before or after the timeout. Can this be replaced with whatever event it is you are waiting for?

Vir-8 replied:
Do you mean incorporating the wait in the base user click functions and such? That might be unwanted as they are common for all utilities

Although we try to avoid timeouts, these ones are very crucial for particular buttons that deal with creating, saving or publishing changes. 

The event we are waiting for is the buttons to be enabled or clickable but without the timeout, the "click" literally never happens for some reason - either it happens way too fast as the button enables or due to transitions, at least on my device. I've only left the ones in that are absolutely necessary.

seanlip replied:
Can we then include the timeout in "wait for enabled/clickable", with a comment/explanation and a pointer to a specific example?

For specific cases with transitions I think it's fine to keep it here though, but in each such "exception case" there should be a comment pointing to the code that specifies the transition duration. Otherwise no one will know why it's 500 as opposed to 1000, 250, etc. and this will make it confusing to fix flakes.

Vir-8 replied:
Sounds good, it's a recurring pattern and after a bit more investigation, seems to always happen when the "click" should happen after a modal closes or opens.

Will make a "wait for clickable" function with the appropriate explanation, thanks.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+  /**
+   * This function checks if the topic with given subtopic and skill is published.
+   */
+  async expectPublishedTopicToBePresent(): Promise<void> {
```
The logic here seems a bit indirect, I would just navigate to the topic viewer page and around it to see what the learner can see?

In general try to avoid anything that does this.page.$eval() innerText stuff -- this is not what an end user would do. Interact exclusively through the actions of an end user. Ditto elsewhere.

Vir-8 replied:
How else would you recommend to verify the contents of the page? Does this.page.eval textContent work?

I'm not sure if this.page.eval itself is a problem since all other utility files follow the same pattern for expectance functions, in order to verify page contents.

seanlip replied:
I think I gave a suggestion in my first paragraph? The thing I don't understand about the current logic is that it seems to be opening the topic editor and looking for a subtopic and skill name; none of this seems to have anything to do with whether the topic is published or not. (Or did I misunderstand the question in your first paragraph?)

For this.page.eval ... I guess that is OK if it can be mapped to the actions of an end user, e.g. look at the first topic in the list and verify that its name is XYZ. I think I was concerned that I couldn't do that with this function, but you're right that it might not be an issue with this.page.$eval itself.

Vir-8 replied:
I'm not really sure what you initially meant by navigating around to see what the learner can see. 

Also I believe I misunderstood the functionality and the phrasing of my comments can be worked upon as well, the expectance function was meant to verify if the topic with the subtopic and skill is published -- and verify that the names of the subtopic and skill are correct, but I overlooked that topics can be saved and don't have to be published for them to exist.

I'll update the functions accordingly - thanks!

seanlip replied:
> I'm not really sure what you initially meant by navigating around to see what the learner can see.

Ah sorry. If you want to check that a topic is published, then go to /learn/math or the relevant topic viewer page and see if some other random user can see it. (I think unpublished topics are not viewable from the learner side.)

But now that Im not quite sure what exactly you're trying to verify, the above might be moot :P

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+  /**
+   * Function for creating a question in the skill editor page.
+   */
+  async createQuestion(): Promise<void> {
```
Many of the other functions start by going to a particular page -- shouldn't this one do so too, or at least assert that the current page is the correct one?

Ditto for other functions -- this seems important for this user because they can do things in quite a lot of places.

Vir-8 replied:
You're right, it currently has to be in flow and has to be called right after creating a skill for it to work. Updated it so that it works regardless of the order.

Other functions all are independent aside from creating chapters and publishing the story, we do not want to refresh the page in the middle of story creation.

seanlip replied:
OK, sounds good, but one note: if you expect function A to be called before function B then either combine those into one function and don't expose the inner ones, or have an assertion at the start of B that the URL points to the correct page.

seanlip reviewed (changes requested):
Thanks @Vir-8! Definitely fine to include the utility for questions. I left some comments, PTAL.

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Improve extracting exploration ID and handle exceptions

Vir-8 committed:
Correct wordings and input texts

Vir-8 committed:
Remove redundant wait and make question creation functionality independent of user flow

StephenYu2018 commented on core/tests/puppeteer-acceptance-tests/spec/curriculum-admin-tests/create-and-publish-topics-and-stories.spec.ts:
```diff
@@ -0,0 +1,65 @@
+        : '';
+
+      await curriculumAdmin.navigateToTopicAndSkillsDashboardPage();
+      await curriculumAdmin.createTopic();
+      await curriculumAdmin.createSubTopic();
+      await curriculumAdmin.createSkill();
+      await curriculumAdmin.createStory();
+      await curriculumAdmin.createChapter(explorationId);
```
Is it possible to pass in objects containing information about user inputs as arguments in these `create...` methods? For example, instead of `createSkill()`, do `createSkill(skill)` where `skill = {description, reviewMaterial, misconceptions, difficulties, questions}` is stored as a constant for this suite.

Vir-8 replied:
Yep, that sounds great, thanks for the suggestion!

StephenYu2018 commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,412 @@
+    await this.page.waitForSelector(
+      `${confirmSkillCreationButton}:not([disabled])`
+    );
+    await this.page.waitForTimeout(500);
```
Ideally we shouldn't be waiting a hard-coded number of seconds to pass as it makes the test more flaky. Either the test doesn't wait long enough for the necessary UI components to load into the correct state which will cause the test to fail, or the test waits much longer than expected and passes. The same test passes on some test runs, while it fails on other test runs, which is what I mean by "flaky tests".

For each `waitForTimeout` statement, consider investigating why the test would fail without the statement. The other `waitFor...` methods such as `waitForSelector` with options `visible` or `hidden` are often much more reliable and it clarifies what state the surrounding UI needs to be in before executing the user action.

Ditto for all the other `waitForTimeout` statements you've added.

Vir-8 replied:
The tests fail in the absence of the timeouts because of modal transitions - when modals open or close. That leads to the click never happening, which is why the timeout is needed and waitFor methods don't suffice. I've only left in the ones that are absolutely necessary for buttons as far as I'm aware.

If I'm not wrong, the transitions were of 0.3 seconds so 0.5 seconds is a safe space for the tests to pass in my opinion.

seanlip replied:
I seem to recall @StephenYu2018 has some code in one of his PRs (see withinModal() in puppeteer-utils.js [here](https://github.com/oppia/oppia/pull/19869/files#diff-94cb0186ac035dfd0c36611e18f5832cd2c04e097de358044814f7e0cf13371fR217)) that introduces a general "do stuff in modal" functionality. Maybe we can reuse that idea here, if this only happens in modals?

StephenYu2018 replied:
@seanlip I still need to merge the PR before the rest of the acceptance test team can access `withinModal()`.

@Vir-8 For now, try something like `waitForSelector(selector, {visible: true})` to wait for the modal to appear before executing user interactions with the modal. Also you could try `waitForSelector(selector, {hidden: true})` to wait for the modal to completely disappear before executing user interactions with UI components behind the modal. Also, these techniques aren't just for modals; you can apply them to any UI element that appears and/or disappears upon user interactions.

EDIT: You can read more about `waitForSelector` options [here](https://pptr.dev/api/puppeteer.waitforselectoroptions).

Vir-8 replied:
@seanlip @StephenYu2018 Thanks for the suggestions, appreciate it!

StephenYu2018 reviewed (changes requested):
2 concerns before merging.

Vir-8 committed:
Add comments explaining timeouts and generalize functionality for timeout event

Vir-8 committed:
Fix opening skill editor before question creation

Vir-8 committed:
Refactor story creation to combine functions following user flow

Vir-8 committed:
Modify functionalities to use objects for user input information

Vir-8 committed:
Fix handling necessary question creation

Vir-8 committed:
Complete topic creation functionality including publishing topic

Vir-8 committed:
Appropriately publish topics, update expectance test and replace timeouts

Vir-8 committed:
Add comments for clarity

Vir-8 committed:
Move helper function to spec file

Vir-8 committed:
Clean code and add comments

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix linting issues

Vir-8 committed:
Minimize flaky behaviour appropriately

Vir-8 committed:
Clean code

Vir-8 committed:
Modify tests to emulate mobile user behaviour

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Adjust utility to publish topics for mobile functionality

StephenYu2018 reviewed (approved):
LGTM!

vojtechjelinek reviewed (approved):
LGTM!

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Remove test timeouts

Vir-8 committed:
Improve saving topic draft flow

Vir-8 committed:
Fix new tab flake and finalize mobile utility functions

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/curriculum-admin-tests/create-and-publish-topics-and-stories.spec.ts:
```diff
@@ -0,0 +1,107 @@
+  let explorationUrl: string | null;
+  let explorationId: string;
+
+  let exploration = {
```
In the top level test file, you shouldn't have all this data and the getExplorationIdFromUrl. The top-level test files should read like a user journey spec and not have technical detail. 

Please move getExplorationIdFromUrl to a suitable place (e.g. a general-purpose url parser utility service or something that the base user can do). For things like createTopic etc. you should pass arguments explicitly (e.g. `createTopic('Test Topic 1', 'test-topic-one')`) and autogenerate the web_fragment and description from the given arguments if they're not that important for the purposes of the test (and if they're important, pass them explicitly as additional arguments).

Vir-8 replied:
I was actually told to pass them as objects but made the changes, done!

seanlip reviewed (changes requested):
Hi @Vir-8, one comment re the top-level test layout, PTAL. Thanks.

Vir-8 committed:
Move exploration ID extraction function to base user utility service

Vir-8 committed:
Remove objects and pass data as explicit arguments

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix linting issues

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+    for (let i = 0; i < questionCount; i++) {
+      await this.createQuestion();
+    }
+    await this.assignSkillToSubtopic();
```
This should be reflected in the function name (e.g. createAndAssignSkillToSubtopic), and the name of the subtopic to assign the skill to should be passed in.

Vir-8 replied:
Done / Changed scope of function

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+  }
+
+  /**
+   * Function for creating a question in the skill editor page.
```
Give a brief description here about the type of question that's being created.

In general, update each docstring so that it's possible to get a clear mental picture of what the function is doing just by reading the docstring.

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+   * We need to generate at least 3 questions for the skill as a
+   * pre-requisite for publishing the topic later.
+   */
+  async createSkill(description: string, questionCount: number): Promise<void> {
```
Explain the prerequisites for this function in the docstring, and check them within the code if possible. For example, if you require the user to be on a particular page at the point the function starts, have a utility function like verifyUserIsOnTopicEditorPage() that checks the URL has the right format and errors if it doesn't.

Ditto for other functions in this file.

Also, one issue I am seeing in this file is that the functions are not consistent about the above. I think either you can require a prerequisite (see above) or navigate to the page directly at the start of the function, as you do here. But if you do the latter then the topic name/ID has to be passed in so that the code can navigate to the right "topic editor" page.

seanlip replied:
A further thought about this -- I was reviewing another acceptance test PR and wondering why I didn't feel the same way about the functions in here: https://github.com/oppia/oppia/pull/19780/files#diff-885139b0d44e4d97ca4c17e8b918ed7a53dcdcc08814a1b3728a625a2c8542c4R92

I think it's because, in the other PR, it is easy to tell what the functions are doing **on the page**. So, if a test attempts e.g. switchToEditorTab() in a context that doesn't make sense, then it's easy to identify that. With createSkill() it is much harder (I can't mentally map that to an action in the UI). 

So just noting that a third alternative to the above is to name the function clearly so that it's possible to easily visualize what it does.

Vir-8 replied:
Done -- changed the scope of the function and improved naming

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+  /**
+   * Function for creating an exploration as a curriculum admin.
+   */
+  async createExploration(state: string, title: string): Promise<string> {
```
"state" is unclear. Do you mean "initialContentText" or something like that? (Also, if that's the case, the argument should probably come after the title.)

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+  /**
+   * Function that opens the topic editor page for the topic created.
+   */
+  async openTopicEditor(): Promise<void> {
```
Doesn't this need to take the topic name as an argument?

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+  /**
+   * Function that opens the skill editor page for the topic created.
+   */
+  async openSkillEditor(): Promise<void> {
```
Ditto, doesn't this need the skill as an argument? Otherwise if we ever write a test that exercises a scenario with > 1 skills, we cannot disambiguate.

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+  /**
+   * Function for assigning a skill to a subtopic in the topic editor page.
+   */
+  async assignSkillToSubtopic(): Promise<void> {
```
I think this needs the skill and subtopic as arguments, logically

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+   * This function checks if the topic has been published successfully,
+   * by verifying the status and the counts in the topic and skills dashboard.
+   */
+  async expectPublishedTopicToBePresent(): Promise<void> {
```
The name of this function is still not clear. When a dev reads the name of the function they should know what actions are being taken to verify, or should be able to infer that. E.g.  expectPublishedTopicToBePresentInMathClassroom() or similar.

Please clarify (and see the previous comment about going to /learn/math or the relevant topic viewer page).

Vir-8 replied:
Hey, if I'm not misunderstanding, classroom creation and then adding a topic to it is a different user story based on this [testing spreadsheet](https://docs.google.com/spreadsheets/d/1O8EHiSAGrG0yoNUBz9E4DIwKNS8Rfsv_ffC4k1WK5jc/edit#gid=1416806242).

Currently, I am verifying the actions by just checking the topic's status in the topics and skills dashboard, and verifying the number of published stories and such. Are we merging the 2 user stories?

seanlip replied:
Ah, I see, sorry, you're right. OK to check via the topics and skills dashboard; in that case maybe have the name be expectTopicsAndSkillsDashboardToShowTopicIsPublished(topicName) or similar.

seanlip reviewed (changes requested):
Thanks @Vir-8! The top-level looks OK so I started looking at the next layer.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+const assignSubtopicButton = '.e2e-test-assign-subtopic';
+
+const skillsTab = 'a.e2e-test-skills-tab';
+const skill = 'a.e2e-test-open-skill-editor';
```
I don't think this is named correctly.

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,703 @@
+const explorationIdElement = 'span.oppia-unique-progress-id';
+
+const topicsTab = 'a.e2e-test-topics-tab';
+const topic = 'a.e2e-test-topic-name';
```
I don't think this is named correctly. This is a topic link that can be clicked on, right? (A "topic" refers to the underlying domain object.)

Vir-8 replied:
Done

Vir-8 committed:
Address comments and improve clarity

Vir-8 committed:
Improve user utility functionality

Vir-8 committed:
Reduce flakiness and improve naming

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix linter issues

Vir-8 committed:
Fix flake and improve naming

Vir-8 commented:
@seanlip I've addressed all your comments and really hope it's better, I'm aware that it's failing on CI but I'm unable to reproduce the error manually and have made a debugging doc at #20184 regarding the same.

I don't see any problems in the code, PTAL meanwhile as the issue might cause extended delays.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/curriculum-admin-tests/create-and-publish-topics-and-stories.spec.ts:
```diff
@@ -0,0 +1,85 @@
+        'Test Subtopic 1',
+        'Test Topic 1'
+      );
+      await curriculumAdmin.addSkillToDiagnostingTestsOfTopic(
```
addSkillToDiagnosticTest(....)

(note: there is 1 diagnostic test for the math classroom, and it covers all topics)

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/curriculum-admin-tests/create-and-publish-topics-and-stories.spec.ts:
```diff
@@ -0,0 +1,85 @@
+        explorationId,
+        'Test Topic 1'
+      );
+      await curriculumAdmin.expectTopicToBePublishedInTopicAndSkillsDashboard(
```
It's TopicsAndSkillsDashboard

Vir-8 replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -0,0 +1,825 @@
+
+export class CurriculumAdmin extends BaseUser {
+  /**
+   * Function for navigating to the topic and skills dashboard page.
```
You can omit "Function for ...", "Function to ..." -- try to document these similar to what we do in Python, start the docstrings with a simple verb.

Vir-8 replied:
Done

seanlip reviewed (approved):
I have barely any concerns with the code, it looks good in general. Thanks! Please fix the minor nits and feel free to get this submitted.

Sorry to hear about the failing test though. In general it would be great to have screenshot functionality so we can take a screenshot at the failure point in GitHub Actions....or a video perhaps. Maybe one thing you can do is to try doing other actions on the page to get a sense of what's there and what's not -- basically try to log as much info as you can about the page to tell you something about its structure at the point the test fails.

Vir-8 committed:
Fix typos and improve comments

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

Vir-8 committed:
Fix merge issues and update suite

Vir-8 committed:
Attempt to fix CI error

Vir-8 committed:
Merge remote-tracking branch 'upstream/develop' into curriculum-admin-topic-management-acceptance-test

jnvtnguyen commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -13,31 +13,829 @@
+  /**
+   * Navigate to the creator dashboard page.
+   */
+  async navigateToCreatorDashboardPage(): Promise<void> {
```
Aren't we duplicating code here? Thought this was apart of exploration editor utils already.

Vir-8 replied:
Done

jnvtnguyen commented on core/tests/puppeteer-acceptance-tests/user-utilities/curriculum-admin-utils.ts:
```diff
@@ -13,31 +13,829 @@
+  /**
+   * Create an exploration as a curriculum admin.
+   */
+  async createAndPublishExploration(
```
Same here, I would expect this to be in the exploration editor utils, unless I'm missing some context.

Vir-8 replied:
We're testing the curriculum admin role here though - not the exploration editor. Curriculum admins can also create and publish explorations. Ditto above, seems to be recently pushed code.

jnvtnguyen replied:
Sorry to clarify AFAIK we are trying to separate user behaviors. Yes a curriculum admin can create and publish explorations but so can every user, we wouldn't want to duplicate this for every user role. We use composition which allows a user to have multiple classes. ATM the exploration editor class is composed by default so any function you put in there can be ran by any created user.

jnvtnguyen replied:
Unless this function has something specific to the curriculum admin role we should separate it into the exploration editor class.

Vir-8 replied:
Ah okay, thanks for explaining, so are you suggesting to create a separate exploration editor user in the spec to deal with exploration creation if I'm not misunderstanding? I didn't quite get what you meant by a user having multiple classes.

jnvtnguyen replied:
No it's fine to have one user just move these functions over to the other file and concat the types.

jnvtnguyen replied:
E.g. when we you declare curriculumAdmin: CurriculumAdmin & ExplorationEditor

Vir-8 replied:
Done, thanks!

Vir-8 committed:
Create explorations from exploration editor utils to avoid code duplication

Vir-8 committed:
Remove unused variables

jnvtnguyen reviewed (approved):
Thanks, LGTM.
